mypy/plugins/enums.py

```diff
@@ -78,6 +78,22 @@ class SomeEnum:
     """
     enum_field_name = _extract_underlying_field_name(ctx.type)
     if enum_field_name is None:
+        # We do not know the enum field name (perhaps it was passed to a function and we only
+        # know that it _is_ a member).  All is not lost however, if we can prove that the all
+        # of the enum members have the same value-type, then it doesn't matter which member
+        # was passed in.  The value-type is still known.
+        if isinstance(ctx.type, Instance):
+            info = ctx.type.type
+            stnodes = (info.get(name) for name in info.names)
+            first_node = next(stnodes, None)
+            if first_node is None:
+                return ctx.default_attr_type
+            first_node_type = first_node.type
+            if all(node is not None and node.type == first_node_type for node in stnodes):
+                underlying_type = get_proper_type(first_node_type)
+                if underlying_type is not None:
+                    return underlying_type
+
         return ctx.default_attr_type
 
     assert isinstance(ctx.type, Instance)
```

Comment:
Below on lines 91-92 is a TODO that appears relevant. Unfortunately it appears hard to do something useful here -- while the default implementations of `auto()` always return ints, the docs specifically suggest that you can override it and return anything you like (the docs show it returning the field name).

So I'm not sure if you're going to be able to do much about that, but I thought I'd point it out as a potential improvement for your PR. It would certainly be nice if we got `int` for the following example:
```
class Truth(Enum):
    false = auto()
    true = auto()
```
As well as for a similar class where the first value is `0` instead of `auto()`.

But you'd have to check whether they're overriding `_generate_next_value_()` in the class.

Reply:
It took a little digging into the mypy guts here, but in the end, getting that working wasn't _too_ bad.  Running locally, the following code:

```py
from enum import Enum, auto
from typing import NamedTuple, List, Any


class AColorVal(NamedTuple):
    c: str


class Color(Enum):
    RED = 1
    BLUE = 2


def print_color_value(color: Color) -> None:
    reveal_type(color.value)


class Color2(Enum):
    RED = (1,)
    BLUE = (2,)


def print_color2_value(color: Color2) -> None:
    reveal_type(color.value)


class Color3(Enum):
    RED = AColorVal("RED")
    BLUE = AColorVal("BLUE")


def print_color3_value(color: Color3) -> None:
    reveal_type(color.value)


class Color4(Enum):
    RED = 10
    BLUE = auto()


def print_color4_value(color: Color4) -> None:
    reveal_type(color.value)


print(Color4.BLUE.value)


class AutoName(Enum):

    @staticmethod
    def _generate_next_value_(name: str, start: int, count: int, last_values: List[Any]) -> str:
        return name

class Color5(AutoName):
    NORTH = auto()
    SOUTH = auto()
    EAST = auto()
    WEST = auto()


def print_color5_value(color: Color5) -> None:
    reveal_type(color.value)


class Color6(Enum):
    RED = auto()
    BLUE = auto()

    def foo(self) -> str:
        return "bar"


def print_color6_value(color: Color6) -> None:
    reveal_type(color.value)
```

Produces the expected output:

```sh
/Users/mgilson/sandbox/foo.py:15: note: Revealed type is 'builtins.int'
/Users/mgilson/sandbox/foo.py:24: note: Revealed type is 'Tuple[builtins.int]'
/Users/mgilson/sandbox/foo.py:33: note: Revealed type is 'Tuple[builtins.str, fallback=foo.AColorVal]'
/Users/mgilson/sandbox/foo.py:42: note: Revealed type is 'builtins.int'
/Users/mgilson/sandbox/foo.py:62: note: Revealed type is 'builtins.str'
/Users/mgilson/sandbox/foo.py:74: note: Revealed type is 'builtins.int'
```

Unfortunately, the tests aren't quite working out (I keep getting complaints about `auto` not having the correct number of arguments.  I'm guessing that there is something interesting going on in the way the fixtures are set up that I haven't completely wrapped my head around (presumably sucking in a bunch of stuff from `typeshed` would slow them down a whole lot).